app/services/course/assessment/answer/auto_grading_service.rb

```diff
@@ -41,18 +41,30 @@ def save!(answer, reattempt)
     end
   end
 
-  # Grades into the given +Course::Assessment::Answer::AutoGrading+ object. This does not do
-  # anything, except mark the answer as having been graded.
-  #
-  # Subclasses should call this implementation after they are done to persist the changes to the
-  # database.
+  # Grades into the given +Course::Assessment::Answer::AutoGrading+ object. This assigns the grade
+  # and makes sure answer is in the correct state.
   #
   # @param [Course::Assessment::Answer] answer The answer to be graded.
   # @return [Course::Assessment::Answer] The graded answer. Note that this answer is not persisted
   #   yet.
   def grade(answer)
-    answer.publish!
-    answer.grader = User.system
+    grade = evaluate(answer)
+    answer.evaluate!
+
+    if answer.question.assessment.autograded?
```

Comment:
Discussed offline, agreed that we can think the conventions for where to place services and models, then refactor for there. This will work for now 👍